enhancements/installer/assisted-installer-in-cluster.md

```diff
@@ -0,0 +1,641 @@
+goals stated above.
+* Run metal3 components on a non-baremetal cluster.
+
+## Proposal
+
+### User Stories
+
+#### Day 2 Add Worker
+
+After deploying a cluster with Assisted Installer, I can add workers from bare
+metal hardware on day two using a similar workflow and tool set. I can either
+obtain the discovery live ISO and use my own methods to boot it, or I can use
+the baremetal-operator to boot the live ISO automatically.
+
+#### Multi-cluster
+
+As a user of Red Hat's multi-cluster mangement tools, I can use the assisted
+installation agent-based workflow to create clusters from a pool of bare metal
+inventory.
+
+### Implementation Details/Notes/Constraints
+
+#### Assisted Installer k8s-native API
+
+The Assisted Installer has a [REST
+API](https://generator.swagger.io/?url=https://raw.githubusercontent.com/openshift/assisted-service/master/swagger.yaml)
+that is available at cloud.redhat.com. The service is implemented with a
+traditional relational database. In order to integrate with OpenShift
+infrastructure management tooling, it is necessary to additionally expose its
+capabilities as Kubernetes-native APIs. They will be implemented as Custom
+Resource Definitions and a controller with the assisted service's capabilities.
+
+For the first deliverable, a local deployment of the backing service will
+include new controllers that will operate to achieve the desired state as
+declared in CRDs. Thus the service will expose both REST and Kubernetes-native
+APIs, and both API layers will call the same set of internal methods. The local
+service will include a database; sqlite if possible, or else postgresql.
+
+    --------------------------
+    |  REST API  |  k8s API  |
+    --------------------------
+    | Assisted Service logic |
+    --------------------------
+        |               |
+        v               v
+      SQL DB        File System
+                      storage
+
+For modest clusters that do not have persistent storage available but need
+day-2 ability to add nodes, the database will need to be reconstructable in
+case its current state is lost. Thus the CRs will be the source of truth for
+the specs, and the controller will upon statup first ensure that the DB
+reflects the desired state in CRs. The source of truth for the status, however,
+is the actual state of the agents running on the hosts being installed and not
+necessarily what was previously recorded.
+
+Agents that are running on hardware and waiting for installation-related
+instructions will continue to communicate with the assisted service using the
+existing REST API. In the future that may transition to using a CRD.
+
+**AgentDrivenInstallationImage**
+This resource represents a discovery image that should be used for booting
+hosts. In the REST API this resource is embedded in the "cluster" resource, but
+for kubernetes-native is more natural for it to be separate.
+
+Spec:
+ * clusterName: string
```

Comment:
Do we anticipate an image existing in a different namespace than the cluster it's related to?

Reply:
@filanov I guess we move to back strings

I guess things like secrets should still be proper references?  For example I could imagine the pull-secret being used by multiple entities in multiple namespaces.

Reply:
Pull secret is tough to use in multiple namespaces as you can't mount in a secret for a pod. The pod would need very broad secret RBAC across the cluster, which is troublesome. 

ACM fights with this as they keep a central index of cloud creds users can use for lots of clusters in different namespaces. @jnpacker was commenting on the challenges keeping these in sync in all namespaces and we floated some ideas, nothing concrete yet.

I'd suggest Secret be LocalReference as well, but if you need to go to secrets in another namespace, I believe best practice is to create your own type instead of using ObjectReference. (from our experience everyone is trying to stay away from it now)